### ISSUE TITLE
read "kid_count failed" ! vw: std::exception

### ISSUE THREAD

**Issue by amit024003** [JohnLangford/vowpal_wabbit#1280]
Hi,

I am running VW on hadoop cluster with spanning tree. I am training model with various feature combinations one after another. For example: 
Feature set: af1, bf2, cf3, df4

1st Training: af1 and bf2 (using -ignore c -ignore d)
2nd Training: bf2 and cf3
3rd training: af1, cf3 and df4

For some combination my job is successful but some it fails with following error:
read "kid_count failed" ! 
vw: std::exception

Please help me to understand the root cause of this problem and resolve this.

**Comment by amit024003** [comment]
./vw -b 24
        --total 666
        --node 000278
        --unique_id 8495324
        --passes 40
        --save_per_pass
        --readable_model automotive_klq_40_0.0001.rmodel
        -d /dev/stdin
        -f automotive_klq_40_0.0001.vwmodel
        --cache_file temp.cache
        --span_server <hostname>
        --loss_function=logistic
        --l1 0.0001  --ignore a  --ignore e  --ignore c  --ignore g
ignoring namespaces beginning with: a e c g 
using l1 regularization = 0.0001
final_regressor = automotive_klq_40_0.0001.vwmodel
Num weight bits = 24
learning rate = 0.5
initial_t = 0
power_t = 0.5
decay_learning_rate = 1
can't open: temp.cache, error = No such file or directory
creating cache_file = temp.cache
Reading datafile = /dev/stdin
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
0.693147   0.693147            1         1.0  -1.0000   0.0000       23
0.662319   0.631491            2         2.0  -1.0000  -0.1274       30
0.596604   0.530888            4         4.0   1.0000   0.9865       77
3.047721   5.498838            8         8.0  -1.0000  21.3127       88
2.439788   1.831855           16        16.0   1.0000   0.2471       76
1.456558   0.473329           32        32.0   1.0000   1.2154       65
0.912245   0.367933           64        64.0   1.0000   1.8645       61
0.502297   0.092348          128       128.0   1.0000   1.8453       86
0.332237   0.162176          256       256.0   1.0000   5.3170       73
0.423071   0.513906          512       512.0  -1.0000  -1.1678       75
0.380144   0.337216         1024      1024.0   1.0000   1.2838       86
0.249440   0.118736         2048      2048.0   1.0000   2.5792       57
0.181177   0.112914         4096      4096.0   1.0000   6.7139       92
0.154970   0.128763         8192      8192.0   1.0000   2.5224       80
0.250720   0.346470        16384     16384.0  -1.0000  -0.1153       57
0.285604   0.320488        32768     32768.0   1.0000  -0.3699       52
connecting to <ip_address> 
wrote unique_id=8495324
wrote total=666
wrote node=278
read ok=1
read kid_count failed!
vw: std::exception

**Comment by arielf** [comment]
Looks like an issue in recovering from a network read error in `vowpalwabbit/allreduce_sockets.cc`

See the source:
[https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/allreduce_sockets.cc#L140](https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/allreduce_sockets.cc#L140)

Clearly there a network issue between the two hosts.
Maybe there should be a exponential backoff retry there before giving up?

**Comment by amit024003** [comment]
I see. So, I can retry several times instead of exiting the process. Good suggestion. Let me do that.

**Comment by arielf** [comment]
I think ideally the C++ code should be improved too so the error actually says what the problem is.

See `man recv` on Linux.

For starters, a better diagnostic should be printed based on the return value, `errno` and `SO_EE_OFFENDER(struct sock_extended_err *)` - Note that this may be linux only (no Windows or even Mac equivalent).

Even if no detail is added, the error message should be improved to clarify the fact that this is a failure to network-read an expected response (which) from one of the members (which one) of the cluster.

**Comment by amit024003** [comment]
So, let me keep it open. So, in future someone can work on it.

**Comment by JohnLangford** [comment]
I updated the error message to be a bit more explicit.  @arielf if you want to do something even better, that's certainly welcome.
